narwhals/_pandas_like/dataframe.py

```diff
@@ -167,16 +167,47 @@ def filter(
 
     def with_columns(
         self,
-        *exprs: IntoPandasExpr | Iterable[IntoPandasExpr],
+        *exprs: IntoPandasExpr,
         **named_exprs: IntoPandasExpr,
     ) -> Self:
+        from narwhals._pandas_like.expr import PandasExpr
+
+        index = self._dataframe.index
         new_series = evaluate_into_exprs(self, *exprs, **named_exprs)
-        df = self._dataframe.assign(
-            **{
-                series.name: validate_dataframe_comparand(self._dataframe.index, series)
-                for series in new_series
-            }
+        # If the inputs are all Expressions which return full columns
+        # (as opposed to scalars), we can use a fast path (concat, instead of assign).
+        # We can't use the fastpath if any input is not an expression (e.g.
+        # if it's a Series) because then we might be changing its flags.
+        # See `test_memmap` for an example of where this is necessary.
```

Comment:
Holy moly that's a great catch! 

I am getting a bit confused on the wording "e.g. if it's a Series", doesn't `validate_dataframe_comparand` check if other is a Series? (I am reading such function for the first time though)

Reply:
great question! so

- expr: IntoPandasExpr. this is something that can be converted to an expression, like:
  - a string (with a column name)
  - a PandasSeries
  - a PandasExpr
- evaluate_into_expr evaluates each input `expr` into a `PandasSeries`
- `validate_dataframe_comparand` returns the underlying `pandas.Series` from each `PandasSeries`, possibly replacing the `Index`, see https://narwhals-dev.github.io/narwhals/other/pandas_index/

At this stage, we're checking whether the input `exprs` are `PandasExpr`, before they have been converted to `PandasSeries`